File: src/VbNet.LanguageServer/Protocol/MessageDispatcher.cs
// JSON-RPC message dispatcher for routing LSP requests and notifications

using System.Collections.Concurrent;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace VbNet.LanguageServer.Protocol;

/// <summary>
/// Dispatches incoming JSON-RPC messages to registered handlers.
/// Parses raw JSON into typed messages and routes based on method name.
/// </summary>
public sealed class MessageDispatcher
{
    private readonly ILogger<MessageDispatcher> _logger;
    private readonly ITransport _transport;

    private readonly Dictionary<string, Func<JsonElement?, CancellationToken, Task<object?>>> _requestHandlers = new();
    private readonly Dictionary<string, Func<JsonElement?, CancellationToken, Task>> _notificationHandlers = new();
    private readonly ConcurrentDictionary<JsonRpcId, CancellationTokenSource> _requestCancellation = new();

    private const string CancelRequestMethod = "$/cancelRequest";

    public MessageDispatcher(ITransport transport, ILogger<MessageDispatcher> logger)
    {
        _transport = transport ?? throw new ArgumentNullException(nameof(transport));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    /// <summary>
    /// Registers a handler for an LSP request method.
    /// </summary>
    public void RegisterRequest<TParams, TResult>(
        string method,
        Func<TParams?, CancellationToken, Task<TResult>> handler)
    {
        _requestHandlers[method] = async (paramsElement, ct) =>
        {
            var parameters = paramsElement.HasValue
                ? JsonSerializer.Deserialize<TParams>(paramsElement.Value.GetRawText(), JsonSerializerOptionsProvider.Options)
                : default;
            return await handler(parameters, ct);
        };
        _logger.LogDebug("Registered request handler for: {Method}", method);
    }

    /// <summary>
    /// Registers a handler for an LSP notification method.
    /// </summary>
    public void RegisterNotification<TParams>(
        string method,
        Func<TParams?, CancellationToken, Task> handler)
    {
        _notificationHandlers[method] = async (paramsElement, ct) =>
        {
            var parameters = paramsElement.HasValue
                ? JsonSerializer.Deserialize<TParams>(paramsElement.Value.GetRawText(), JsonSerializerOptionsProvider.Options)
                : default;
            await handler(parameters, ct);
        };
        _logger.LogDebug("Registered notification handler for: {Method}", method);
    }

    /// <summary>
    /// Registers a handler for a notification with no parameters.
    /// </summary>
    public void RegisterNotification(string method, Func<CancellationToken, Task> handler)
    {
        _notificationHandlers[method] = async (_, ct) => await handler(ct);
        _logger.LogDebug("Registered notification handler for: {Method}", method);
    }

    /// <summary>
    /// Sends a notification to the client.
    /// </summary>
    public async Task SendNotificationAsync<TParams>(string method, TParams parameters, CancellationToken cancellationToken = default)
    {
        var notification = new JsonRpcNotification
        {
            Method = method,
            Params = JsonSerializer.SerializeToElement(parameters, JsonSerializerOptionsProvider.Options)
        };

        var json = JsonSerializer.Serialize(notification, JsonSerializerOptionsProvider.Options);
        await _transport.WriteMessageAsync(json, cancellationToken);
        _logger.LogDebug("Sent notification: {Method}", method);
    }

    /// <summary>
    /// Sends a notification with no parameters to the client.
    /// </summary>
    public async Task SendNotificationAsync(string method, CancellationToken cancellationToken = default)
    {
        var notification = new JsonRpcNotification { Method = method };
        var json = JsonSerializer.Serialize(notification, JsonSerializerOptionsProvider.Options);
        await _transport.WriteMessageAsync(json, cancellationToken);
        _logger.LogDebug("Sent notification: {Method}", method);
    }

    /// <summary>
    /// Starts the message processing loop.
    /// </summary>
    public async Task RunAsync(CancellationToken cancellationToken)
    {
        _logger.LogInformation("Message dispatcher started");
        var inFlightTasks = new ConcurrentBag<Task>();

        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                var message = await _transport.ReadMessageAsync(cancellationToken);
                if (message == null)
                {
                    _logger.LogInformation("Transport closed, stopping message dispatcher");
                    break;
                }

                var task = ProcessMessageWithHandlingAsync(message, cancellationToken);
                inFlightTasks.Add(task);
            }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
            {
                _logger.LogDebug("Message dispatcher cancelled");
                break;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error processing message");
                // Continue processing other messages
            }
        }

        await AwaitInFlightTasksAsync(inFlightTasks).ConfigureAwait(false);
        _logger.LogInformation("Message dispatcher stopped");
    }

    private async Task ProcessMessageWithHandlingAsync(string messageJson, CancellationToken cancellationToken)
    {
        try
        {
            await ProcessMessageAsync(messageJson, cancellationToken).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            // Ignore cancellations from shutdown.
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing message");
        }
    }

    private static async Task AwaitInFlightTasksAsync(ConcurrentBag<Task> tasks)
    {
        if (tasks.IsEmpty)
        {
            return;
        }

        try
        {
            await Task.WhenAll(tasks).ConfigureAwait(false);
        }
        catch
        {
            // Individual tasks log their own errors.
        }
    }

    private async Task ProcessMessageAsync(string messageJson, CancellationToken cancellationToken)
    {
        JsonDocument document;
        try
        {
            document = JsonDocument.Parse(messageJson);
        }
        catch (JsonException ex)
        {
            _logger.LogError(ex, "Failed to parse JSON-RPC message");
            return;
        }

        using (document)
        {
            var root = document.RootElement;

            // Check if this is a request (has id) or notification (no id)
            var hasId = root.TryGetProperty("id", out var idElement);
            var hasMethod = root.TryGetProperty("method", out var methodElement);

            if (!hasMethod)
            {
                _logger.LogWarning("Received message without method property");
                return;
            }

            var method = methodElement.GetString()!;
            var paramsElement = root.TryGetProperty("params", out var p) ? p : (JsonElement?)null;

            if (string.Equals(method, CancelRequestMethod, StringComparison.Ordinal))
            {
                await HandleCancelRequestAsync(paramsElement).ConfigureAwait(false);
                return;
            }

            if (hasId)
            {
                // This is a request
                var id = ParseId(idElement);
                await HandleRequestAsync(id, method, paramsElement, cancellationToken);
            }
            else
            {
                // This is a notification
                await HandleNotificationAsync(method, paramsElement, cancellationToken);
            }
        }
    }

    private async Task HandleRequestAsync(JsonRpcId id, string method, JsonElement? paramsElement, CancellationToken cancellationToken)
    {
        _logger.LogDebug("Handling request: {Method} (id: {Id})", method, id);

        JsonRpcResponse response;
        CancellationTokenSource? requestCts = null;
        CancellationTokenSource? linkedCts = null;
        var requestCancellationToken = cancellationToken;

        if (!id.IsNull)
        {
            requestCts = new CancellationTokenSource();
            if (!_requestCancellation.TryAdd(id, requestCts))
            {
                requestCts.Dispose();
                requestCts = null;
            }
        }

        if (requestCts != null)
        {
            linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, requestCts.Token);
            requestCancellationToken = linkedCts.Token;
        }

        if (_requestHandlers.TryGetValue(method, out var handler))
        {
            try
            {
                var result = await handler(paramsElement, requestCancellationToken);
                response = JsonRpcResponse.Success(id, result);
            }
            catch (OperationCanceledException)
            {
                response = JsonRpcResponse.CreateError(id, JsonRpcErrorCodes.RequestCancelled, "Request cancelled");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error handling request: {Method}", method);
                response = JsonRpcResponse.CreateError(id, JsonRpcErrorCodes.InternalError, ex.Message);
            }
        }
        else
        {
            _logger.LogWarning("No handler registered for request: {Method}", method);
            response = JsonRpcResponse.CreateError(id, JsonRpcErrorCodes.MethodNotFound, $"Method not found: {method}");
        }

        try
        {
            var json = JsonSerializer.Serialize(response, JsonSerializerOptionsProvider.Options);
            await _transport.WriteMessageAsync(json, cancellationToken);
        }
        finally
        {
            if (!id.IsNull && _requestCancellation.TryRemove(id, out var cts))
            {
                cts.Dispose();
            }

            linkedCts?.Dispose();
        }
    }

    private async Task HandleNotificationAsync(string method, JsonElement? paramsElement, CancellationToken cancellationToken)
    {
        _logger.LogDebug("Handling notification: {Method}", method);

        if (_notificationHandlers.TryGetValue(method, out var handler))
        {
            try
            {
                await handler(paramsElement, cancellationToken);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error handling notification: {Method}", method);
            }
        }
        else
        {
            _logger.LogTrace("No handler registered for notification: {Method}", method);
        }
    }

    private Task HandleCancelRequestAsync(JsonElement? paramsElement)
    {
        if (!paramsElement.HasValue)
        {
            return Task.CompletedTask;
        }

        try
        {
            var cancelParams = JsonSerializer.Deserialize<CancelParams>(
                paramsElement.Value.GetRawText(),
                JsonSerializerOptionsProvider.Options);
            if (cancelParams == null || cancelParams.Id.IsNull)
            {
                return Task.CompletedTask;
            }

            if (_requestCancellation.TryRemove(cancelParams.Id, out var cts))
            {
                cts.Cancel();
                cts.Dispose();
                _logger.LogDebug("Cancelled request: {Id}", cancelParams.Id);
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to process $/cancelRequest");
        }

        return Task.CompletedTask;
    }

    private static JsonRpcId ParseId(JsonElement element)
    {
        return element.ValueKind switch
        {
            JsonValueKind.String => new JsonRpcId(element.GetString()!),
            JsonValueKind.Number => new JsonRpcId(element.GetInt64()),
            _ => default
        };
    }
}

File: src/VbNet.LanguageServer.Vb/Protocol/MessageDispatcher.vb
' JSON-RPC message dispatcher for routing LSP requests and notifications

Imports System.Collections.Concurrent
Imports System.Text.Json
Imports Microsoft.Extensions.Logging

Namespace Protocol

    ''' <summary>
    ''' Dispatches incoming JSON-RPC messages to registered handlers.
    ''' Parses raw JSON into typed messages and routes based on method name.
    ''' </summary>
    Public NotInheritable Class MessageDispatcher
        Private ReadOnly _logger As ILogger(Of MessageDispatcher)
        Private ReadOnly _transport As ITransport

        Private ReadOnly _requestHandlers As Dictionary(Of String, Func(Of JsonElement?, CancellationToken, Task(Of Object))) =
            New Dictionary(Of String, Func(Of JsonElement?, CancellationToken, Task(Of Object)))()
        Private ReadOnly _notificationHandlers As Dictionary(Of String, Func(Of JsonElement?, CancellationToken, Task)) =
            New Dictionary(Of String, Func(Of JsonElement?, CancellationToken, Task))()
        Private ReadOnly _requestCancellation As ConcurrentDictionary(Of JsonRpcId, CancellationTokenSource) =
            New ConcurrentDictionary(Of JsonRpcId, CancellationTokenSource)()
        Private ReadOnly _requestTasks As ConcurrentBag(Of Task) = New ConcurrentBag(Of Task)()

        Private Const CancelRequestMethod As String = "$/cancelRequest"

        Public Sub New(transport As ITransport, logger As ILogger(Of MessageDispatcher))
            If transport Is Nothing Then
                Throw New ArgumentNullException(NameOf(transport))
            End If
            If logger Is Nothing Then
                Throw New ArgumentNullException(NameOf(logger))
            End If

            _transport = transport
            _logger = logger
        End Sub

        ''' <summary>
        ''' Registers a handler for an LSP request method.
        ''' </summary>
        Public Sub RegisterRequest(Of TParams, TResult)(method As String, handler As Func(Of TParams, CancellationToken, Task(Of TResult)))
            _requestHandlers(method) = Async Function(paramsElement, ct)
                                           Dim parameters As TParams = Nothing
                                           If paramsElement.HasValue Then
                                               parameters = JsonSerializer.Deserialize(Of TParams)(paramsElement.Value.GetRawText(), JsonSerializerOptionsProvider.Options)
                                           End If

                                           Dim result = Await handler(parameters, ct).ConfigureAwait(False)
                                           Return CType(result, Object)
                                       End Function
            _logger.LogDebug("Registered request handler for: {Method}", method)
        End Sub

        ''' <summary>
        ''' Registers a handler for an LSP notification method.
        ''' </summary>
        Public Sub RegisterNotification(Of TParams)(method As String, handler As Func(Of TParams, CancellationToken, Task))
            _notificationHandlers(method) = Async Function(paramsElement, ct)
                                                Dim parameters As TParams = Nothing
                                                If paramsElement.HasValue Then
                                                    parameters = JsonSerializer.Deserialize(Of TParams)(paramsElement.Value.GetRawText(), JsonSerializerOptionsProvider.Options)
                                                End If

                                                Await handler(parameters, ct).ConfigureAwait(False)
                                            End Function
            _logger.LogDebug("Registered notification handler for: {Method}", method)
        End Sub

        ''' <summary>
        ''' Registers a handler for a notification with no parameters.
        ''' </summary>
        Public Sub RegisterNotification(method As String, handler As Func(Of CancellationToken, Task))
            _notificationHandlers(method) = Async Function(ignoredParams, ct)
                                                Await handler(ct).ConfigureAwait(False)
                                            End Function
            _logger.LogDebug("Registered notification handler for: {Method}", method)
        End Sub

        ''' <summary>
        ''' Sends a notification to the client.
        ''' </summary>
        Public Async Function SendNotificationAsync(Of TParams)(method As String, parameters As TParams, Optional cancellationToken As CancellationToken = Nothing) As Task
            Dim notification = New JsonRpcNotification With {
                .Method = method,
                .Params = JsonSerializer.SerializeToElement(parameters, JsonSerializerOptionsProvider.Options)
            }

            Dim json = JsonSerializer.Serialize(notification, JsonSerializerOptionsProvider.Options)
            Await _transport.WriteMessageAsync(json, cancellationToken).ConfigureAwait(False)
            _logger.LogDebug("Sent notification: {Method}", method)
        End Function

        ''' <summary>
        ''' Sends a notification with no parameters to the client.
        ''' </summary>
        Public Async Function SendNotificationAsync(method As String, Optional cancellationToken As CancellationToken = Nothing) As Task
            Dim notification = New JsonRpcNotification With {.Method = method}
            Dim json = JsonSerializer.Serialize(notification, JsonSerializerOptionsProvider.Options)
            Await _transport.WriteMessageAsync(json, cancellationToken).ConfigureAwait(False)
            _logger.LogDebug("Sent notification: {Method}", method)
        End Function

        ''' <summary>
        ''' Starts the message processing loop.
        ''' </summary>
        Public Async Function RunAsync(cancellationToken As CancellationToken) As Task
            _logger.LogInformation("Message dispatcher started")

            While Not cancellationToken.IsCancellationRequested
                Try
                    Dim message = Await _transport.ReadMessageAsync(cancellationToken).ConfigureAwait(False)
                    If message Is Nothing Then
                        _logger.LogInformation("Transport closed, stopping message dispatcher")
                        Exit While
                    End If

                    Await ProcessMessageWithHandlingAsync(message, cancellationToken).ConfigureAwait(False)
                Catch ex As OperationCanceledException When cancellationToken.IsCancellationRequested
                    _logger.LogDebug("Message dispatcher cancelled")
                    Exit While
                Catch ex As Exception
                    _logger.LogError(ex, "Error processing message")
                End Try
            End While

            _logger.LogInformation("Message dispatcher stopped")
        End Function

        Private Async Function ProcessMessageWithHandlingAsync(messageJson As String, cancellationToken As CancellationToken) As Task
            Try
                Await ProcessMessageAsync(messageJson, cancellationToken).ConfigureAwait(False)
            Catch ex As OperationCanceledException When cancellationToken.IsCancellationRequested
                ' Ignore cancellations from shutdown.
            Catch ex As Exception
                _logger.LogError(ex, "Error processing message")
            End Try
        End Function

        Private Async Function ProcessMessageAsync(messageJson As String, cancellationToken As CancellationToken) As Task
            Dim document As JsonDocument
            Try
                document = JsonDocument.Parse(messageJson)
            Catch ex As JsonException
                _logger.LogError(ex, "Failed to parse JSON-RPC message")
                Return
            End Try

            Using document
                Dim root = document.RootElement

                ' Check if this is a request (has id) or notification (no id)
                Dim idElement As JsonElement
                Dim hasId = root.TryGetProperty("id", idElement)
                Dim methodElement As JsonElement
                Dim hasMethod = root.TryGetProperty("method", methodElement)

                If Not hasMethod Then
                    _logger.LogWarning("Received message without method property")
                    Return
                End If

                Dim method = methodElement.GetString()
                Dim paramsElement As JsonElement? = Nothing
                Dim paramsValue As JsonElement
                If root.TryGetProperty("params", paramsValue) Then
                    paramsElement = paramsValue
                End If

                If String.Equals(method, CancelRequestMethod, StringComparison.Ordinal) Then
                    Await HandleCancelRequestAsync(paramsElement).ConfigureAwait(False)
                    Return
                End If

                If hasId Then
                    Dim id = ParseId(idElement)
                    Dim task = HandleRequestAsync(id, method, paramsElement, cancellationToken)
                    _requestTasks.Add(task)
                    Dim ignored = task.ContinueWith(
                        Sub(t)
                            _logger.LogError(t.Exception, "Request handler failed for: {Method} (id: {Id})", method, id)
                        End Sub,
                        CancellationToken.None,
                        TaskContinuationOptions.OnlyOnFaulted,
                        TaskScheduler.Default)
                Else
                    Await HandleNotificationAsync(method, paramsElement, cancellationToken).ConfigureAwait(False)
                End If
            End Using
        End Function

        Private Async Function HandleRequestAsync(id As JsonRpcId, method As String, paramsElement As JsonElement?, cancellationToken As CancellationToken) As Task
            _logger.LogDebug("Handling request: {Method} (id: {Id})", method, id)

            Dim response As JsonRpcResponse
            Dim requestCts As CancellationTokenSource = Nothing
            Dim linkedCts As CancellationTokenSource = Nothing
            Dim requestCancellationToken = cancellationToken

            If Not id.IsNull Then
                requestCts = New CancellationTokenSource()
                If Not _requestCancellation.TryAdd(id, requestCts) Then
                    requestCts.Dispose()
                    requestCts = Nothing
                End If
            End If

            If requestCts IsNot Nothing Then
                linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, requestCts.Token)
                requestCancellationToken = linkedCts.Token
            End If

            Dim handler As Func(Of JsonElement?, CancellationToken, Task(Of Object)) = Nothing
            If _requestHandlers.TryGetValue(method, handler) Then
                Try
                    Dim result = Await handler(paramsElement, requestCancellationToken).ConfigureAwait(False)
                    response = JsonRpcResponse.Success(id, result)
                Catch ex As OperationCanceledException
                    response = JsonRpcResponse.CreateError(id, JsonRpcErrorCodes.RequestCancelled, "Request cancelled")
                Catch ex As Exception
                    _logger.LogError(ex, "Error handling request: {Method}", method)
                    response = JsonRpcResponse.CreateError(id, JsonRpcErrorCodes.InternalError, ex.Message)
                End Try
            Else
                _logger.LogWarning("No handler registered for request: {Method}", method)
                response = JsonRpcResponse.CreateError(id, JsonRpcErrorCodes.MethodNotFound, $"Method not found: {method}")
            End If

            Try
                Dim json = JsonSerializer.Serialize(response, JsonSerializerOptionsProvider.Options)
                Await _transport.WriteMessageAsync(json, cancellationToken).ConfigureAwait(False)
            Finally
                If Not id.IsNull Then
                    Dim removed As CancellationTokenSource = Nothing
                    If _requestCancellation.TryRemove(id, removed) Then
                        removed.Dispose()
                    End If
                End If

                If linkedCts IsNot Nothing Then
                    linkedCts.Dispose()
                End If
            End Try
        End Function

        Private Async Function HandleNotificationAsync(method As String, paramsElement As JsonElement?, cancellationToken As CancellationToken) As Task
            _logger.LogDebug("Handling notification: {Method}", method)

            Dim handler As Func(Of JsonElement?, CancellationToken, Task) = Nothing
            If _notificationHandlers.TryGetValue(method, handler) Then
                Try
                    Await handler(paramsElement, cancellationToken).ConfigureAwait(False)
                Catch ex As Exception
                    _logger.LogError(ex, "Error handling notification: {Method}", method)
                End Try
            Else
                _logger.LogTrace("No handler registered for notification: {Method}", method)
            End If
        End Function

        Private Function HandleCancelRequestAsync(paramsElement As JsonElement?) As Task
            If Not paramsElement.HasValue Then
                Return Task.CompletedTask
            End If

            Try
                Dim idElement As JsonElement
                If paramsElement.Value.ValueKind = JsonValueKind.Object AndAlso paramsElement.Value.TryGetProperty("id", idElement) Then
                    Dim id = ParseId(idElement)
                    If id.IsNull Then
                        Return Task.CompletedTask
                    End If

                    Dim cts As CancellationTokenSource = Nothing
                    If _requestCancellation.TryRemove(id, cts) Then
                        cts.Cancel()
                        cts.Dispose()
                        _logger.LogDebug("Cancelled request: {Id}", id)
                    End If
                End If
            Catch ex As Exception
                _logger.LogWarning(ex, "Failed to process $/cancelRequest")
            End Try

            Return Task.CompletedTask
        End Function

        Private Shared Function ParseId(element As JsonElement) As JsonRpcId
            Select Case element.ValueKind
                Case JsonValueKind.String
                    Return New JsonRpcId(element.GetString())
                Case JsonValueKind.Number
                    Return New JsonRpcId(element.GetInt64())
                Case Else
                    Return New JsonRpcId()
            End Select
        End Function
    End Class

End Namespace
